File: web-component/javascript/src/api/AutoGenerateAppTokenApi.js
(function(root, factory) {
    if (typeof define === 'function' && define.amd) {
      // AMD. Register as an anonymous module.
      define(['ApiClient', 'model/AppToken','AuthApi'], factory);
    } else if (typeof module === 'object' && module.exports) {
      // CommonJS-like environments that support module.exports, like Node.
      module.exports = factory(require('../ApiClient'), require('../model/AppToken'), require('./AuthApi'));
    } else {
      // Browser globals (root is window)
      if (!root.HydrogenAdminApi) {
        root.HydrogenAdminApi = {};
      }
      root.HydrogenAdminApi.AutoGenerateAppTokenApi = factory(root.HydrogenAdminApi.ApiClient, root.HydrogenAdminApi.AppToken, root.HydrogenAdminApi.AuthApi);
    }
  }(this, function(ApiClient, AppToken, AuthApi) {
    'use strict';
  
    /**
     * AppToken service.
     * @module api/GetAppTokenApi
     * @version 1.3.0
     */
  
    /**
     * Constructs a new GetAppTokenApi. 
     * @alias module:api/GetAppTokenApi
     * @class
     * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
     * default to {@link module:ApiClient#instance} if unspecified.
     */
    var exports = function(apiClient, authApi) {
      this.apiClient = apiClient || ApiClient.instance;
      this.authApi = authApi || new AuthApi();
      
      this.getAppTokenUsingGET = function(appTokenConfig, callback) {

        
        let template  = "<tag app-token='##app_token##' ##attrib_map##></tag>";
        var postBody = null;
  
        // verify the required parameter 'appName' is set
        if (appTokenConfig.appName === undefined || appTokenConfig.appName === null) {
          throw new Error("Missing the required parameter 'appName' when calling getAppTokenUsingGET");
        }

        if (appTokenConfig.basePath === undefined || appTokenConfig.basePath === null) {
          throw new Error("Missing the required parameter 'basePath' when calling getAppTokenUsingGET");
        }

        this.apiClient.basePath = appTokenConfig.basePath;

  
        var pathParams = {
        };
        var queryParams = {
        };
        var collectionQueryParams = {
          'app_name': {
            value: appTokenConfig.appName,
            collectionFormat: 'multi'
          },
        };
        var headerParams = {
        };
        var formParams = {
        };
  
        var authNames = ['oauth2'];
        var contentTypes = [];
        var accepts = ['*/*'];
        var returnType = [AppToken];

        var oauth2 = this.apiClient.authentications['oauth2'];

       
        let finalAttribMap = [];
        if(appTokenConfig.attribMap){
          finalAttribMap = [...new Set(appTokenConfig.attribMap.map(it => it.name + "='"+it.value+"'"))];
        }

        async function createUsingPostClientCredentials (apiClient, authApi) {
          // Token Generation for grant_type = client_credentials
          const req = authApi.createUsingPostClientCredentials({
            'grant_type': 'client_credentials',
            'client_id': appTokenConfig.clientId,
            'client_secret': appTokenConfig.clientSecret
          }, null);

          return new Promise(function(resolve, reject) {
              req
              .then(response => {
                let data = null;
                try {
                    data = apiClient.deserialize(response, Object);
                    if (apiClient.enableCookies && typeof window === 'undefined') {
                        apiClient.agent.saveCookies(response);
                    }
                } catch (err) {
                    console.log(err);
                    reject(err);
                }

                resolve(data)
              })
              .catch(err => {
                console.error(err);
                reject(err);
             });
            });
        }

        async function createUsingPostPassword (apiClient, authApi, username, password) {
          // Token Generation for grant_type = password
          const req = authApi.createUsingPostPassword({
            'grant_type': 'password',
            'client_id': appTokenConfig.clientId,
            'client_secret': appTokenConfig.clientSecret,
            'username': username,
            'password': password
          }, null);

          return new Promise(function(resolve, reject) {
              req
              .then(response => {
                let data = null;
                try {
                    data = apiClient.deserialize(response, Object);
                    if (apiClient.enableCookies && typeof window === 'undefined') {
                        apiClient.agent.saveCookies(response);
                    }
                } catch (err) {
                    console.log(err);
                    reject(err);
                }

                resolve(data)
              })
              .catch(err => {
                console.error(err);
                reject(err);
             });
            });
        }

        async function getAppToken (apiClient, appName) {
          const req = apiClient.callApi(
                    '/app_token?app_name=' + appName, 'GET',
                    pathParams, {}, {}, headerParams, formParams, postBody,
                    authNames, contentTypes, accepts, [AppToken], null
                  );

          return new Promise(function(resolve, reject) {
              req
              .then(response => {
                let data = null;
                try {
                    data = apiClient.deserialize(response, Object);
                    if (apiClient.enableCookies && typeof window === 'undefined') {
                        apiClient.agent.saveCookies(response);
                    }
                } catch (err) {
                    console.log(err);
                    reject(err);
                }

                resolve(data)
              })
              .catch(err => {
                console.error(err);
                reject(err);
             });
            });
        }

        try {
        (async () => {
          const tokenData = await createUsingPostClientCredentials(this.apiClient, this.authApi).catch(e=>{
            callback(e, null, null)
          });
          oauth2.accessToken = tokenData.access_token;
            let response = [];
              for (var appConfig in appTokenConfig.appName) {
                var app =  appTokenConfig.appName[0];
                let item = {};
                  if(app.auth_type && app.auth_type.toLowerCase() === "client_credentials"){
                    oauth2.accessToken = tokenData.access_token;

                  }else if(app.auth_type && app.auth_type.toLowerCase() === "password_credentials"){
                    oauth2.accessToken = appTokenConfig.userAccessToken;
                    if(appTokenConfig.isCredsPassed){
                      const passwordTokenData = await createUsingPostPassword(this.apiClient, this.authApi, appTokenConfig.username, appTokenConfig.password).catch(e=>{
                        callback(e, null, null)
                      });
                      oauth2.accessToken = passwordTokenData.access_token;
                    }
                  }
                  const appTokenData = await getAppToken(this.apiClient, app.app_name).catch(e=>{
                    callback(e, null, null)
                  });
                  const appTokenValue = appTokenData && appTokenData.length>0 ? appTokenData[0].app_token : '';
                  const tagValue = app.app_name.toLowerCase().replace(/_/g, '-');
                  const fillTemplateValue = template.replace(/tag/g, tagValue)
                                            .replace(/##app_token##/g, appTokenValue)
                                            .replace(/##attrib_map##/g, finalAttribMap!=null ?  finalAttribMap.join(' ') : '');
                  
                  item[app.app_name] = appTokenValue;
                  if(appTokenConfig.isEmbed){
                    item[app.app_name] = fillTemplateValue;
                  }
                  response.push(item);
              }
            callback(null, response, response) ;
        })();
      }catch(e){
        callback(e, null, null);
      }
     }
  };
  
    return exports;
  }));
  

File: web-component/dotnet/src/main/CsharpDotNet2/IO/Swagger/Api/AutoGenerateAppTokenApi.cs
using System;
using System.Collections.Generic;
using RestSharp;
using IO.Swagger.Client;
using IO.Swagger.Model;

namespace IO.Swagger.Api
{
    /// <summary>
    /// Represents a collection of functions to interact with the API endpoints
    /// </summary>
    public interface IAutoGenerateAppTokenApi
    {
        /// <summary>
        /// getAppToken 
        /// </summary>
        List<Dictionary<String, String>> GetAppTokenUsingGET ();
    }
  
    /// <summary>
    /// Represents a collection of functions to interact with the API endpoints
    /// </summary>
    public class AutoGenerateAppTokenApi : IAutoGenerateAppTokenApi
    {
        private AppTokenConfig appTokenConfig;
        private String template = "<tag app-token='##app_token##' ##attrib_map##></tag>";
        /// <summary>
        /// Initializes a new instance of the <see cref="AutoGenerateAppTokenApi"/> class.
        /// </summary>
        /// <param name="apiClient"> an instance of ApiClient (optional)</param>
        /// <returns></returns>
        public AutoGenerateAppTokenApi(AppTokenConfig appTokenConfig)
        {
            this.ApiClient = Configuration.DefaultApiClient; 
            this.ApiClient.BasePath = appTokenConfig.basePath;
            this.appTokenConfig = appTokenConfig;
        }

    
        /// <summary>
        /// Sets the base path of the API client.
        /// </summary>
        /// <param name="basePath">The base path</param>
        /// <value>The base path</value>
        public void SetBasePath(String basePath)
        {
            this.ApiClient.BasePath = basePath;
        }
    
        /// <summary>
        /// Gets the base path of the API client.
        /// </summary>
        /// <param name="basePath">The base path</param>
        /// <value>The base path</value>
        public String GetBasePath(String basePath)
        {
            return this.ApiClient.BasePath;
        }
    
        /// <summary>
        /// Gets or sets the API client.
        /// </summary>
        /// <value>An instance of the ApiClient</value>
        public ApiClient ApiClient {get; set;}
    
        /// <summary>
        /// getAppToken 
        /// </summary>
        /// <param name="appName">app_name</param> 
        /// <param name="tenantName">tenant-name</param> 
        /// <returns>List&lt;AppToken&gt;</returns>            
        public List<AppToken> GetAppTokenUsingGET (List<string> appName, string tenantName, string token)
        {
            const string AUTHORIZATION = "Authorization";
            const string BEARER = "Bearer ";
            // verify the required parameter 'appName' is set
            if (appName == null) throw new ApiException(400, "Missing required parameter 'appName' when calling GetAppTokenUsingGET");
            this.ApiClient.BasePath = this.appTokenConfig.basePath;
            String baseUri = this.appTokenConfig.basePath.Substring(0, appTokenConfig.basePath.IndexOf(".com") + 4);
            this.ApiClient.BasePath = baseUri;
           
            var path = "/admin/v1/app_token";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
             if (appName != null) queryParams.Add("app_name", ApiClient.ParameterToString(appName)); // query parameter
             if (tenantName != null) headerParams.Add("tenant-name", ApiClient.ParameterToString(tenantName)); // header parameter
             
             headerParams[AUTHORIZATION] = BEARER + token;
            // authentication setting, if any
            String[] authSettings = new String[] { "oauth2" };
           
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.GET, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling GetAppTokenUsingGET: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling GetAppTokenUsingGET: " + response.ErrorMessage, response.ErrorMessage);
    
            return (List<AppToken>) ApiClient.Deserialize(response.Content, typeof(List<AppToken>), response.Headers);
        }

        public List<Dictionary<String, String>> GetAppTokenUsingGET ()
        {
             List<Dictionary<String, String>> localVarResponse = GetAppTokenUsingGETAsyncWithHttpInfo();
             return localVarResponse;
        }

          public List<Dictionary<String, String>> GetAppTokenUsingGETAsyncWithHttpInfo ()
        {
            List<Dictionary<String, String>> response = new List<Dictionary<String, String>>();
            String token = null;
            AuthApiClient authApiClient = new AuthApiClient(this.ApiClient);
            String clientCredentialToken = authApiClient.createClientCredential(appTokenConfig.clientId, appTokenConfig.clientSecret);
            token = clientCredentialToken;

                foreach (AppConfig app in appTokenConfig.appNames) {
                    if (app.authType != null && app.authType.ToLower()=="client_credentials") {
                        token = clientCredentialToken;
                    } else if (app.authType != null && app.authType.ToLower() == "password_credentials") {
                        token = appTokenConfig.userAccessToken;
                        if (appTokenConfig.isCredsPassed) {
                            String passwordCredentialToken = authApiClient.createPasswordCredential(appTokenConfig.clientId, appTokenConfig.clientSecret, appTokenConfig.username, appTokenConfig.password);
                            token = passwordCredentialToken;
                        }
                    }

                    List<AppToken> appTokens = GetAppTokenUsingGET(new List<String> {app.appName}, null, token);
                    if (appTokens != null) {
                        String appTokenValue = appTokens != null && appTokens.Count > 0 ? appTokens[0]._AppToken : "";
                        String tagValue = app.appName.ToLower().Replace("_", "-");
                        String fillTemplateValue = template.Replace("tag", tagValue)
                                .Replace("##app_token##", appTokenValue)
                                .Replace("##attrib_map##", getAttribMapList(appTokenConfig) != null ? string.Join(" ", getAttribMapList(appTokenConfig)) : "");

                        Dictionary<String, String> appMap = new Dictionary<String, String>();
                        

                        if (appTokenConfig.isEmbed) {
                            appMap.Add(app.appName, fillTemplateValue);
                        }else{
                            appMap.Add(app.appName, appTokenValue);
                        }
                        response.Add(appMap);
                    }
                }
            return response;
        }
        private String[] getAttribMapList(AppTokenConfig appTokenConfig) {
            List<TagAttributes> attributes = appTokenConfig.attributes;
            List<String> finalAttributes = new List<String>();

            foreach (TagAttributes tagAttributes in attributes?? new List<TagAttributes>()) {
                finalAttributes.Add(tagAttributes.name + "='" + tagAttributes.value + "'");
            }
            return finalAttributes.ToArray();
        }
    
    }
}
